server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -197,117 +212,202 @@ public void clusterChanged(final ClusterChangedEvent event) {
+            for (final ObjectCursor<IndexMetaData> cursor : state.metaData().indices().values()) {
+                routingTableBuilder.addAsRecovery(cursor.value);
+            }
+            // start with 0 based versions for routing table
+            routingTableBuilder.version(0);
+            return ClusterState.builder(state).routingTable(routingTableBuilder.build()).build();
+        }
+
+        static ClusterState removeStateNotRecoveredBlock(final ClusterState state) {
+            return ClusterState.builder(state)
+                    .blocks(ClusterBlocks.builder()
+                            .blocks(state.blocks()).removeGlobalBlock(STATE_NOT_RECOVERED_BLOCK).build())
+                    .build();
+        }
+
+        static ClusterState mixCurrentStateAndRecoveredState(final ClusterState currentState, final ClusterState recoveredState) {
+            assert currentState.metaData().indices().isEmpty();
+
+            final ClusterBlocks.Builder blocks = ClusterBlocks.builder()
+                    .blocks(currentState.blocks())
+                    .blocks(recoveredState.blocks());
+
+            final MetaData.Builder metaDataBuilder = MetaData.builder(recoveredState.metaData());
+            // automatically generate a UID for the metadata if we need to
+            metaDataBuilder.generateClusterUuidIfNeeded();
+
+            for (final IndexMetaData indexMetaData : recoveredState.metaData()) {
+                metaDataBuilder.put(indexMetaData, false);
+            }
+
+            return ClusterState.builder(currentState)
+                    .blocks(blocks)
+                    .metaData(metaDataBuilder)
+                    .build();
+        }
+
+    }
+
+    class RecoverStateUpdateTask extends ClusterStateUpdateTask {
+
+        @Override
+        public ClusterState execute(final ClusterState currentState) {
+            final ClusterState newState = Function.<ClusterState>identity()
+                    .andThen(state -> Updaters.closeBadIndices(state, indicesService))
+                    .andThen(state -> Updaters.upgradeAndArchiveUnknownOrInvalidSettings(state, clusterService))
+                    .andThen(Updaters::recoverClusterBlocks)
+                    .andThen(Updaters::updateRoutingTable)
+                    .andThen(Updaters::removeStateNotRecoveredBlock)
+                    .apply(currentState);
+
+            return allocationService.reroute(newState, "state recovered");
+        }
+
+        @Override
+        public void clusterStateProcessed(final String source, final ClusterState oldState, final ClusterState newState) {
+            logger.info("recovered [{}] indices into cluster_state", newState.metaData().indices().size());
+        }
+
+        @Override
+        public void onFailure(final String source, final Exception e) {
+            logger.error(() -> new ParameterizedMessage("unexpected failure during [{}]", source), e);
```

Comment:
we log already in `GatewayService.this.onFailure`. Why this extra logging here? This `onFailure` method could possibly be called because we failed to publish the cluster state, e.g. because of another concurrent election being triggered and another node becoming master. Certainly not a reason to log at error level. The error logging should probably only be for the case where closing bad indices or upgrading / archiving bad settings would horribly fail?

Reply:
https://github.com/elastic/elasticsearch/pull/36013/commits/86597fcefa93541bb214311e9db8c7634ee75d0b